# **Data Cleaning - Unemployment.xlsx - UnemploymentMedianIncome.csv**

# **Summary of Data Cleaning**
* GFS, GLDAS and population datasets were merged into one
* The final dataset is collected from 2015 April 1 to December June 30 with the below dates missing due to data absence:</br>
{
'2015-04-10', '2015-04-11', '2015-04-12', '2015-04-13', '2015-04-14', '2015-04-15', '2015-04-16', '2015-04-17', '2015-04-18', '2015-04-19', '2016-12-31', '2017-12-31', '2018-12-31', '2019-12-31', '2020-12-31', '2021-12-15', '2021-12-16', '2021-12-17', '2021-12-18', '2021-12-19', '2021-12-20', '2021-12-21', '2021-12-22', '2021-12-23', '2021-12-24', '2021-12-25', '2021-12-26', '2021-12-27', '2021-12-28', '2021-12-29', '2021-12-30', '2021-12-31', '2022-12-31'
}
* 'SWE_tavg', 'SnowDepth_tavg', 'Qsb_tavg', 'Qsm_tavg', 'EvapSnow_tavg' were dropped
* A new feature **'water_availability'** is created and use it as a target variable (For details, see below **Feature Engineering** section)
* The dataset was split into train and test datasets: train --> 2015 to 2021, test --> 2022 & 2023
* Save them in train_bhopal.csv and test_bhopal.csv respectively
* train_bhopal.csv: 2435 rows, 28 columns
* test_bhopal.csv: 545 rows, 28 columns 

**Data Source:**
###### * https://developers.google.com/earth-engine/datasets/catalog/NOAA_GFS0P25
###### * https://developers.google.com/earth-engine/datasets/catalog/NASA_GLDAS_V022_CLSM_G025_DA1D


# **Import Modules**

In [1]:
#### Import the libraries needed
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
from pathlib import Path
import os
import glob

import matplotlib.pyplot as plt
from matplotlib import rcParams
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

import warnings
from statsmodels.tools.sm_exceptions import ConvergenceWarning
warnings.simplefilter('ignore', ConvergenceWarning)
warnings.filterwarnings('ignore')
%matplotlib notebook
%matplotlib inline

# **Set Environment**

In [2]:
import plotly.io as pio
pio.renderers.default = "vscode"

In [3]:
# Set up directory
working_directory = Path.cwd()
# src = working_directory.parents[1]
# data directory
raw_data_directory = working_directory / 'data' / 'raw'
processed_data_directory = working_directory / 'data' / 'processed'
final_data_directory = working_directory / 'data' / 'final'

In [4]:
# Set pd.options to add slide bars
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', 100)

In [5]:
# Set default title color
plt.style.use('fivethirtyeight')

# set default plt figure size
plt.rcParams["figure.figsize"] = [10, 5]
# suptitle
plt.rcParams["figure.titlesize"] = 22
plt.rcParams["figure.titleweight"] = "bold"
plt.rcParams['text.color'] = '#333333'
# title
plt.rcParams["axes.titlesize"] = 16
plt.rcParams["axes.titleweight"] = "bold"
plt.rcParams["axes.titlelocation"] = "left"
plt.rcParams['axes.titlecolor'] = '#333333'
# label
plt.rcParams["axes.labelsize"] = 12
plt.rcParams["axes.labelweight"] = "bold"
plt.rcParams['axes.labelcolor'] = '#333333'
# spines
plt.rcParams["axes.spines.bottom"] = True
plt.rcParams["axes.spines.left"] = True
plt.rcParams["axes.spines.top"] = False
plt.rcParams["axes.spines.right"] = False
# tick
plt.rcParams['xtick.color'] = "#333333"
plt.rcParams['ytick.color'] = "#333333"
# line width
plt.rcParams['lines.linewidth'] = 1

# **Load Unemployment.xlsx - UnemploymentMedianIncome.csv**
* 3277 rows and 100 columns
* Aggregated by the country(1), state(52), and county level

* Remove ',' and convert from string to float data type
* Dataset includes Port Rico --> delete
* Create a variable 'is_county'
* Remove "County" and the state abbrebiation Ex. Autauga County, AL --> Autauga
* Change the column name from "Area_Name" to "county"
* 149 rows 

* Data Dictionary: UrbanInfluenceCodes2013.xls, Ruralurbancontinuumcodes2023

In [160]:
# Convert the 'date' columns to datetime data type
file = os.path.join(raw_data_directory, 'Unemployment.xlsx - UnemploymentMedianIncome.csv')
unemployment_df = pd.read_csv(file, skiprows=4)

In [127]:
unemployment_df.shape

(3277, 100)

In [128]:
unemployment_df.head()

FIPS_Code State           Area_Name  Rural_Urban_Continuum_Code_2013  \
0          0    US       United States                              NaN   
1       1000    AL             Alabama                              NaN   
2       1001    AL  Autauga County, AL                              2.0   
3       1003    AL  Baldwin County, AL                              3.0   
4       1005    AL  Barbour County, AL                              6.0   

   Urban_Influence_Code_2013  Metro_2013 Civilian_labor_force_2000  \
0                        NaN         NaN               142,601,576   
1                        NaN         NaN                 2,147,173   
2                        2.0         1.0                    21,861   
3                        2.0         1.0                    69,979   
4                        6.0         0.0                    11,449   

  Employed_2000 Unemployed_2000  Unemployment_rate_2000  \
0   136,904,853       5,696,723                     4.0   
1     2,047,731          99,442                     4.6   
2        20,971             890                     4.1   
3        67,370           2,609                     3.7   
4        10,812             637                     5.6   

  Civilian_labor_force_2001 Employed_2001 Unemployed_2001  \
0               143,786,537   136,977,996       6,808,541   
1                 2,128,027     2,017,467         110,560   
2                    22,081        21,166             915   
3                    69,569        66,545           3,024   
4                    11,324        10,468             856   

   Unemployment_rate_2001 Civilian_labor_force_2002 Employed_2002  \
0                     4.7               144,839,298   136,455,783   
1                     5.2                 2,112,621     1,987,633   
2                     4.1                    22,161        21,096   
3                     4.3                    69,379        65,881   
4                     7.6                    11,006        10,154   

  Unemployed_2002  Unemployment_rate_2002 Civilian_labor_force_2003  \
0       8,383,515                     5.8               145,660,094   
1         124,988                     5.9                 2,128,668   
2           1,065                     4.8                    22,695   
3           3,498                     5.0                    72,598   
4             852                     7.7                    11,019   

  Employed_2003 Unemployed_2003  Unemployment_rate_2003  \
0   136,944,522       8,715,572                     6.0   
1     2,001,568         127,100                     6.0   
2        21,557           1,138                     5.0   
3        69,010           3,588                     4.9   
4        10,241             778                     7.1   

  Civilian_labor_force_2004 Employed_2004 Unemployed_2004  \
0               146,724,795   138,613,904       8,110,891   
1                 2,138,306     2,018,783         119,523   
2                    23,241        22,146           1,095   
3                    74,843        71,061           3,782   
4                    10,639         9,884             755   

   Unemployment_rate_2004 Civilian_labor_force_2005 Employed_2005  \
0                     5.5               148,597,241   141,000,912   
1                     5.6                 2,140,356     2,045,234   
2                     4.7                    23,887        22,986   
3                     5.1                    76,608        73,581   
4                     7.1                    10,730        10,114   

  Unemployed_2005  Unemployment_rate_2005 Civilian_labor_force_2006  \
0       7,596,329                     5.1               150,707,773   
1          95,122                     4.4                 2,170,007   
2             901                     3.8                    24,425   
3           3,027                     4.0                    79,806   
4             616                     5.7                    10,713   

  E

In [129]:
unemployment_df.tail()

FIPS_Code State                Area_Name  \
3272      72145    PR  Vega Baja Municipio, PR   
3273      72147    PR    Vieques Municipio, PR   
3274      72149    PR   Villalba Municipio, PR   
3275      72151    PR    Yabucoa Municipio, PR   
3276      72153    PR      Yauco Municipio, PR   

      Rural_Urban_Continuum_Code_2013  Urban_Influence_Code_2013  Metro_2013  \
3272                              1.0                        1.0         1.0   
3273                              7.0                       12.0         0.0   
3274                              2.0                        2.0         1.0   
3275                              1.0                        1.0         1.0   
3276                              2.0                        2.0         1.0   

     Civilian_labor_force_2000 Employed_2000 Unemployed_2000  \
3272                    19,805        17,607           2,198   
3273                     2,441         2,120             321   
3274                     7,636         6,648             988   
3275                    10,613         9,005           1,608   
3276                    14,664        12,871           1,793   

      Unemployment_rate_2000 Civilian_labor_force_2001 Employed_2001  \
3272                    11.1                    19,554        17,138   
3273                    13.2                     2,323         1,999   
3274                    12.9                     7,564         6,546   
3275                    15.2                    10,751         8,746   
3276                    12.2                    14,416        12,475   

     Unemployed_2001  Unemployment_rate_2001 Civilian_labor_force_2002  \
3272           2,416                    12.4                    20,219   
3273             324                    13.9                     2,625   
3274           1,018                    13.5                     8,100   
3275           2,005                    18.6                    11,187   
3276           1,941                    13.5                    15,116   

     Employed_2002 Unemployed_2002  Unemployment_rate_2002  \
3272        17,614           2,605                    12.9   
3273         2,159             466                    17.8   
3274         6,820           1,280                    15.8   
3275         8,961           2,226                    19.9   
3276        13,006           2,110                    14.0   

     Civilian_labor_force_2003 Employed_2003 Unemployed_2003  \
3272                    20,789        18,031           2,758   
3273                     2,664         2,234             430   
3274                     8,393         7,134           1,259   
3275                    11,393         9,162           2,231   
3276                    15,161        12,974           2,187   

      Unemployment_rate_2003 Civilian_labor_force_2004 Employed_2004  \
3272                    13.3                    20,607        18,218   
3273                    16.1                     2,554         2,237   
3274                    15.0                     8,622         7,522   
3275                    19.6                    11,170         9,244   
3276                    14.4                    15,446        13,493   

     Unemployed_2004  Unemployment_rate_2004 Civilian_labor_force_2005  \
3272           2,389                    11.6                    20,930   
3273             317                    12.4                     3,062   
3274           1,100                    12.8                     9,506   
3275           1,926                    17.2                    11,247   
3276           1,953                    12.6                    16,695   

     Employed_2005 Unemployed_2005  Unemployment_rate_2005  \
3272        18,296           2,634                    12.6   
3273         2,725             337                    11.0   
3274         8,112           1,394                    14.7   
3275         9,272           1,975                    17.6   
3276        14,

In [130]:
unemployment_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3277 entries, 0 to 3276
Data columns (total 100 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   FIPS_Code                                  3277 non-null   int64  
 1   State                                      3277 non-null   object 
 2   Area_Name                                  3277 non-null   object 
 3   Rural_Urban_Continuum_Code_2013            3219 non-null   float64
 4   Urban_Influence_Code_2013                  3219 non-null   float64
 5   Metro_2013                                 3224 non-null   float64
 6   Civilian_labor_force_2000                  3270 non-null   object 
 7   Employed_2000                              3270 non-null   object 
 8   Unemployed_2000                            3270 non-null   object 
 9   Unemployment_rate_2000                     3270 non-null   float64
 10  Civilian_labor_force_20

In [159]:
# The number of rows that contains NaN at least one
unemployment_df.isna().any(axis=1).sum()

149

In [131]:
unemployment_df.isna().sum()

FIPS_Code                                     0
State                                         0
Area_Name                                     0
Rural_Urban_Continuum_Code_2013              58
Urban_Influence_Code_2013                    58
Metro_2013                                   53
Civilian_labor_force_2000                     7
Employed_2000                                 7
Unemployed_2000                               7
Unemployment_rate_2000                        7
Civilian_labor_force_2001                     7
Employed_2001                                 7
Unemployed_2001                               7
Unemployment_rate_2001                        7
Civilian_labor_force_2002                     7
Employed_2002                                 7
Unemployed_2002                               7
Unemployment_rate_2002                        7
Civilian_labor_force_2003                     7
Employed_2003                                 7
Unemployed_2003                         

In [132]:
unemployment_df.describe().T

count          mean           std  \
FIPS_Code                                  3277.0  31318.964297  16332.539506   
Rural_Urban_Continuum_Code_2013            3219.0      4.938490      2.724553   
Urban_Influence_Code_2013                  3219.0      5.189811      3.506942   
Metro_2013                                 3224.0      0.382754      0.486135   
Unemployment_rate_2000                     3270.0      4.540581      2.042105   
Unemployment_rate_2001                     3270.0      5.233119      2.228461   
Unemployment_rate_2002                     3270.0      5.981193      2.344970   
Unemployment_rate_2003                     3270.0      6.226391      2.329511   
Unemployment_rate_2004                     3270.0      5.879235      2.114957   
Unemployment_rate_2005                     3263.0      5.671713      2.224239   
Unemployment_rate_2006                     3263.0      5.152253      2.101063   
Unemployment_rate_2007                     3270.0      5.106116      2.150555   
Unemployment_rate_2008                     3270.0      6.032722      2.415970   
Unemployment_rate_2009                     3270.0      9.285688      3.484467   
Unemployment_rate_2010                     3272.0      9.568185      3.490076   
Unemployment_rate_2011                     3272.0      8.909719      3.366212   
Unemployment_rate_2012                     3272.0      8.029004      3.169628   
Unemployment_rate_2013                     3272.0      7.580134      3.131038   
Unemployment_rate_2014                     3272.0      6.473869      2.866920   
Unemployment_rate_2015                     3272.0      5.728759      2.459258   
Unemployment_rate_2016                     3272.0      5.427078      2.353035   
Unemployment_rate_2017                     3272.0      4.788661      2.166696   
Unemployment_rate_2018                     3272.0      4.273319      1.868245   
Unemployment_rate_2019                     3272.0      4.099328      1.780749   
Unemployment_rate_2020                     3194.0      6.775485      2.317718   
Unemployment_rate_2021                     3273.0      4.758478      1.853382   
Unemployment_rate_2022                     3273.0      3.677269      1.363705   
Med_HH_Income_Percent_of_State_Total_2021  3193.0     89.538146     19.638529   

                                            min      25%      50%        75%  \
FIPS_Code                                   0.0  19019.0  30017.0  46099.000   
Rural_Urban_Continuum_Code_2013             1.0      2.0      6.0      7.000   
Urban_Influence_Code_2013                   1.0      2.0      5.0      8.000   
Metro_2013                                  0.0      0.0      0.0      1.000   
Unemployment_rate_2000                      1.3      3.2      4.1      5.300   
Unemployment_rate_2001                      1.6      3.8      4.8      6.100   
Unemployment_rate_2002                      1.6      4.5      5.6      6.900   
Unemployment_rate_2003                      1.9      4.7      5.8      7.300   
Unemployment_rate_2004                      1.6      4.5      5.5      6.700   
Unemployment_rate_2005                      2.0      4.2      5.2      6.500   
Unemployment_rate_2006                      1.6      3.8      4.8      5.900   
Unemployment_rate_2007                      1.4      3.7      4.7      5.900   
Unemployment_rate_2008                      1.3      4.4      5.7      7.200   
Unemployment_rate_2009                      2.0      6.9      8.9     11.200   
Unemployment_rate_2010                      2.0      7.3      9.3     11.425   
Unemployment_rate_2011                      1.4      6.7      8.6     10.600   
Unemployment_rate_2012                      1.1      5.9      7.7      9.600   
Unemployment_rate_2013                      1.2      5.5      7.2      9.000   
Unemployment_rate_2014                      1.2      4.7      6.1      7.600   
Unemployment_rate_2015                      1.8      4.2      5.3      6.600   
Unemployment_r

In [133]:
unemployment_df.describe(include='object').T

count unique                   top freq
State                         3277     53                    TX  255
Area_Name                     3277   3276  District of Columbia    2
Civilian_labor_force_2000     3270   3136                 2,986    3
Employed_2000                 3270   3128                 3,902    3
Unemployed_2000               3270   1749                    57   10
Civilian_labor_force_2001     3270   3127                16,954    3
Employed_2001                 3270   3118                 2,447    3
Unemployed_2001               3270   1891                   196   11
Civilian_labor_force_2002     3270   3141                 5,677    4
Employed_2002                 3270   3123                 4,353    3
Unemployed_2002               3270   1993                    42    8
Civilian_labor_force_2003     3270   3137                 3,909    3
Employed_2003                 3270   3141                 3,290    4
Unemployed_2003               3270   1994                    51    8
Civilian_labor_force_2004     3270   3143                 4,406    3
Employed_2004                 3270   3129                 2,496    3
Unemployed_2004               3270   1980                    65    9
Civilian_labor_force_2005     3263   3136                 8,640    3
Employed_2005                 3263   3136                 4,235    3
Unemployed_2005               3263   1966                   289   10
Civilian_labor_force_2006     3263   3138                 4,836    3
Employed_2006                 3263   3121                 7,084    3
Unemployed_2006               3263   1914                   169   11
Civilian_labor_force_2007     3270   3117                 3,961    3
Employed_2007                 3270   3136                 3,240    3
Unemployed_2007               3270   1889                    75    9
Civilian_labor_force_2008     3270   3145                 5,449    4
Employed_2008                 3270   3115                 1,721    4
Unemployed_2008               3270   2040                    49    8
Civilian_labor_force_2009     3270   3147                 4,641    3
Employed_2009                 3270   3122                 7,297    3
Unemployed_2009               3270   2317                    57    9
Civilian_labor_force_2010     3272   3125                 2,955    3
Employed_2010                 3272   3119                 6,920    3
Unemployed_2010               3272   2327                   782    7
Civilian_labor_force_2011     3272   3129                 2,109    3
Employed_2011                 3272   3130                 3,269    3
Unemployed_2011               3272   2283                   456    7
Civilian_labor_force_2012     3272   3141                 9,678    3
Employed_2012                 3272   3116                 4,025    3
Unemployed_2012               3272   2211                    54    8
Civilian_labor_force_2013     3272   3121                 5,056    3
Employed_2013                 3272   3127                 1,199    3
Unemployed_2013               3272   2188                    37    8
Civilian_labor_force_2014     3272   3140                 8,077    3
Employed_2014                 3272   3130                 3,346    3
Unemployed_2014               3272   2027                    39    9
Civilian_labor_force_2015     3272   3118                 2,746    3
Employed_2015                 3272   3138                 1,283    4
Unemployed_2015               3272   1935                    29    9
Civilian_labor_force_2016     3272   3131                 4,643    4
Employed_2016                 3272   3141                 5,089    3
Unemployed_2016               3272   1929                    40    9
Civilian_labor_force_2017     3272   3133                 2,848    3
Employed_2017                 3272   3136                 7,597    4
Unemployed_2017               3272   1825                   366    9
Civilian_labor_force_2018     3272   3131                 1,665    3
Employed_2018              

## **Data Cleanup and Preparation**

* Remove ',' and convert from string to float data type

In [134]:
# Convert the datatype from object to float
object_columns = unemployment_df.select_dtypes(include=['object']).columns
# Remove 'State' and 'Area_Name'
object_columns = object_columns[2:]

In [135]:
# Remove "," to convert string to number
for col in object_columns:
    unemployment_df[col] = unemployment_df[col].str.replace(",", "")

unemployment_df[object_columns] = unemployment_df[object_columns].astype(float)

* Delete row whose 'State' is US --> delete 1 row
* Delete rows whose 'State' is Port Rico --> delete 79 rows

In [136]:
# Remove the country level
unemployment_df = unemployment_df[1:]

# Remove Port Rico, 79 rows
unemployment_df = unemployment_df[unemployment_df['State'] != "PR"].reset_index(drop=True)

* Create a variable 'is_state' --> If 'Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013', 'Metro_2013' are all null, it's a state
* Create 'county' by removing " County" and " Parish", and the state abbrebiation from "Area_name": Ex. Autauga County, AL --> Autauga
* "Area_Name" in Alaska contains many area names ('Census Area', 'Borough', Borough/municipality, 'city/Census Area', 'city/Borough') --> Keep as is
* Delete 'Area_Name'

In [137]:
# Remove the state from Area_Name
county = []
for i in range(len(unemployment_df)):
    state = unemployment_df.loc[i, 'State']
    area = unemployment_df.loc[i, 'Area_Name'].split(f", {state}")[0]
    county.append(area)

In [138]:
# Create state flag
cols = ['Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013', 'Metro_2013']
is_state = unemployment_df[cols].isna().all(axis=1)
unemployment_df['is_state'] = is_state

# Extract only county
unemployment_df['county'] = county
unemployment_df['county'] = unemployment_df['county'].str.replace(" County", "")
unemployment_df['county'] = unemployment_df['county'].str.replace(" Parish", "")

# Drop 'Area_Name'
unemployment_df.drop(['Area_Name'], axis=1, inplace=True)

* Change the column order starts from 'State', 'county', and 'is_county'

In [139]:
# Chnage column order
cols2 = unemployment_df.columns
cols2 = cols2[2:-2] # from 'Rural_Urban_Continuum_Code_2013' to 'Med_HH_Income_Percent_of_State_Total_2021'
cols1 = ['FIPS_Code', 'State', 'county', 'is_state']
cols1.extend(cols2) 
unemployment_df = unemployment_df[cols1]

## **Missing Value Analysis**

* 51 state rows
* State rows all nulls in 'Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013', 'Metro_2013'

In [140]:
# Check null values at the state level
state_unemployment_df = unemployment_df[unemployment_df['is_state']]
num_state_rows = len(state_unemployment_df)
cols_w_missing_values = state_unemployment_df.columns[state_unemployment_df.isna().sum() > 0].values
print(f"THe number of state rows: {num_state_rows}")
print(f"Columns have null: {cols_w_missing_values}")
print()
print("These columns have all null values:")
print(state_unemployment_df[cols_w_missing_values].isna().all())
print()
# Verify there is no missing values in other columns
cols_wo_missing_values = list(set(state_unemployment_df.columns).difference(set(cols_w_missing_values)))
num_missing_values = state_unemployment_df[cols_wo_missing_values].isna().any().sum()
print(f"Number of null in other columns: {num_missing_values}")

THe number of state rows: 51
Columns have null: ['Rural_Urban_Continuum_Code_2013' 'Urban_Influence_Code_2013'
 'Metro_2013']

These columns have all null values:
Rural_Urban_Continuum_Code_2013    True
Urban_Influence_Code_2013          True
Metro_2013                         True
dtype: bool

Number of null in other columns: 0


* 'Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013', 'Metro_2013' --> -1.0

In [141]:
# Impute 'Rural_Urban_Continuum_Code_2013' and 'Urban_Influence_Code_2013'
cols_rural_urban_metro = ['Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013', 'Metro_2013']
idx = (unemployment_df['is_state']) & (unemployment_df.isna().sum(axis=1)>0)
unemployment_df.loc[idx, cols_rural_urban_metro] = -1.0

In [142]:
# Check null values at the State level
unemployment_df[unemployment_df['is_state']].isna().sum().sum()

0

#### **Alaska**
* These counties in AK have null in 'Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013'
* 'Chugach Census Area', 'Copper River Census Area', 'Prince of Wales-Outer Ketchikan Census Area', 'Skagway-Hoonah-Angoon Census Area', 'Wrangell-Petersburg Census Area'
* These areas have been split and combined with other areas. Thus, they didn't exist in 2013. That's the reason the values are missing
* 'Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013' --> -1.0
#### **Luisiana**
* Unemployment-related features will be imputed with 0
* These parishes in AL have null in 2005 and 2006 unemployment-related features. This may be caused by harricane Katrina hit these counties in 2005
* 'Jefferson', 'Orleans', 'Plaquemines', 'St. Bernard', 'St. Charles', 'St. John the Baptist', 'St. Tammany'
* 'Civilian_labor_force_2005', 'Civilian_labor_force_2006', 'Unemployment_rate_2005', 'Unemployment_rate_2006' will be imputed the data obtained from homefacts.com
* Data Source: https://www.homefacts.com/unemployment/Louisiana/

|Parish|labor force 2005|labor force 2006|unemployment rate 2005|unemployment rate 2006|
|---|---|---|---|---|
|Jefferson|433336|436528|3.35|1.94|
|Orleans|409287|397096|4.23|2.81|
|Plaquemines|23481|23662|3.37|2.04|
|St. Bernard|51230|46948|3.77|2.41|
|St. Charles|53019|52798|3.56|2.0|
|St. John the Baptist|19486|16480|4.33|2.86|
|St. Tammany|96332|84484|2.94|1.70|

* https://www.homefacts.com/unemployment/Louisiana/
* LOUISIANA LABOR FORCE DIVERSITY DATA 2006
* Data Srouce: https://www.laworks.net/Downloads/Employment/AffirmativeActionPublication_2006.pdf
* Data Srouce: https://www.laworks.net/downloads/employment/affirmativeactionpublication_2007.pdf
* LOUISIANA LABOR FORCE DIVERSITY DATA 2007

In [143]:
# Check null at the county level
unemployment_df[~(unemployment_df['is_state']) & (unemployment_df.isna().sum(axis=1)>0)]

FIPS_Code State                                       county  is_state  \
74         2063    AK                          Chugach Census Area     False   
75         2066    AK                     Copper River Census Area     False   
80         2105    AK                    Hoonah-Angoon Census Area     False   
91         2195    AK                           Petersburg Borough     False   
92         2198    AK            Prince of Wales-Hyder Census Area     False   
93         2201    AK  Prince of Wales-Outer Ketchikan Census Area     False   
95         2230    AK                         Skagway Municipality     False   
96         2232    AK            Skagway-Hoonah-Angoon Census Area     False   
98         2261    AK                   Valdez-Cordova Census Area     False   
99         2275    AK                        Wrangell Borough/city     False   
100        2280    AK              Wrangell-Petersburg Census Area     False   
1161      22051    LA                                    Jefferson     False   
1171      22071    LA                                      Orleans     False   
1173      22075    LA                                  Plaquemines     False   
1179      22087    LA                                  St. Bernard     False   
1180      22089    LA                                  St. Charles     False   
1183      22095    LA                         St. John the Baptist     False   
1187      22103    LA                                  St. Tammany     False   

      Rural_Urban_Continuum_Code_2013  Urban_Influence_Code_2013  Metro_2013  \
74                                NaN                        NaN         0.0   
75                                NaN                        NaN         0.0   
80                                9.0                       10.0         0.0   
91                                9.0                       11.0         0.0   
92                                9.0                       10.0         0.0   
93                                NaN                        NaN         0.0   
95                                9.0                       12.0         0.0   
96                                NaN                        NaN         0.0   
98                                9.0                       11.0         0.0   
99                                9.0                       12.0         0.0   
100                               NaN                        NaN         0.0   
1161                              1.0                        1.0         1.0   
1171                              1.0                        1.0         1.0   
1173                              1.0                        1.0         1.0   
1179                              1.0                        1.0         1.0   
1180                              1.0                        1.0         1.0   
1183                              1.0                        1.0         1.0   
1187                              1.0                        1.0         1.0   

      Civilian_labor_force_2000  Employed_2000  Unemployed_2000  \
74                          NaN            NaN              NaN   
75                          NaN            NaN              NaN   
80                          NaN            NaN              NaN   
91                          NaN            NaN              NaN   
92                          NaN            NaN              NaN   
93                       2666.0         2289.0            377.0   
95                          NaN            NaN              NaN   
96                       1830.0         1650.0            180.0   
98                       5226.0         4830.0            396.0   
99                          NaN            NaN              NaN   
100                      3077.0         2780.0            297.0   
1161                   231068.0       220742.0          10326.0   
1171                   210183.0       199041.0          11142.0   
1173                    10

In [144]:
cols_rural_urban = ['Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013']
cols_unemployment = unemployment_df.columns[7:] # columns after 'Metro_2013'

* 'Rural_Urban_Continuum_Code_2013', 'Urban_Influence_Code_2013' --> -1.0

In [145]:
# Impute 'Rural_Urban_Continuum_Code_2013' and 'Urban_Influence_Code_2013'
idx = (unemployment_df['State'] == 'AK') & ~(unemployment_df['is_state']) & (unemployment_df[cols_rural_urban].isna().sum(axis=1)>0)
unemployment_df.loc[idx, cols_rural_urban] = -1.0

* Unemployment-related features will be imputed with 0

In [146]:
# Impute 'Rural_Urban_Continuum_Code_2013' and 'Urban_Influence_Code_2013'
idx = (unemployment_df['State'] == 'AK') & ~(unemployment_df['is_state']) & (unemployment_df[cols_unemployment].isna().sum(axis=1)>0)
unemployment_df.loc[idx, cols_unemployment] = 0

* 'Civilian_labor_force_2005', 'Civilian_labor_force_2006', 'Unemployment_rate_2005', 'Unemployment_rate_2006' will be imputed

In [147]:
# Load the missing data collected from homefacts.com
file = os.path.join(raw_data_directory, 'LA_missing_data.csv')
la_missing_data =  pd.read_csv(file)

In [148]:
la_missing_data

Parish  Civilian_labor_force_2005  Civilian_labor_force_2006  \
0             Jefferson                     433336                     436528   
1               Orleans                     409287                     397096   
2           Plaquemines                      23481                      23662   
3           St. Bernard                      51230                      46948   
4           St. Charles                      53019                      52798   
5  St. John the Baptist                      19486                      16480   
6           St. Tammany                      96332                      84484   

   Unemployment_rate_2005  Unemployment_rate_2006  
0                    5.03                    3.88  
1                    6.35                    5.62  
2                    5.59                    4.08  
3                    5.65                    4.82  
4                    5.34                    4.00  
5                    6.50                    5.72  
6                    4.41                    3.40

In [149]:
# Create featires
la_missing_data['Unemployed_2005'] = round(la_missing_data['Civilian_labor_force_2005'] * la_missing_data['Unemployment_rate_2005']/100)
la_missing_data['Employed_2005'] = la_missing_data['Civilian_labor_force_2005'] - la_missing_data['Unemployed_2005']
la_missing_data['Unemployed_2006'] = round(la_missing_data['Civilian_labor_force_2006'] * la_missing_data['Unemployment_rate_2006']/100)
la_missing_data['Employed_2006'] = la_missing_data['Civilian_labor_force_2006'] - la_missing_data['Unemployed_2006']

In [150]:
parish = la_missing_data['Parish']
cols_unemployment_2005_2006 = la_missing_data.columns[1:]

In [151]:
for i in range(len(la_missing_data)):
    county = la_missing_data['Parish'][i]
    for col in la_missing_data.columns[1:]:
        val = la_missing_data.loc[i, col]
        
        idx = (unemployment_df['State'] == 'LA') & (unemployment_df['county'] == county)
        unemployment_df.loc[idx, col] = val

In [152]:
# Verify no more null values
unemployment_df[(unemployment_df['State'].isin(['AK', 'LA'])) & ~(unemployment_df['is_state'])].isna().sum().sum()

0

In [153]:
# Check the final dataset
unemployment_df.head()

FIPS_Code State   county  is_state  Rural_Urban_Continuum_Code_2013  \
0       1000    AL  Alabama      True                             -1.0   
1       1001    AL  Autauga     False                              2.0   
2       1003    AL  Baldwin     False                              3.0   
3       1005    AL  Barbour     False                              6.0   
4       1007    AL     Bibb     False                              1.0   

   Urban_Influence_Code_2013  Metro_2013  Civilian_labor_force_2000  \
0                       -1.0        -1.0                  2147173.0   
1                        2.0         1.0                    21861.0   
2                        2.0         1.0                    69979.0   
3                        6.0         0.0                    11449.0   
4                        1.0         1.0                     8623.0   

   Employed_2000  Unemployed_2000  Unemployment_rate_2000  \
0      2047731.0          99442.0                     4.6   
1        20971.0            890.0                     4.1   
2        67370.0           2609.0                     3.7   
3        10812.0            637.0                     5.6   
4         8160.0            463.0                     5.4   

   Civilian_labor_force_2001  Employed_2001  Unemployed_2001  \
0                  2128027.0      2017467.0         110560.0   
1                    22081.0        21166.0            915.0   
2                    69569.0        66545.0           3024.0   
3                    11324.0        10468.0            856.0   
4                     9134.0         8513.0            621.0   

   Unemployment_rate_2001  Civilian_labor_force_2002  Employed_2002  \
0                     5.2                  2112621.0      1987633.0   
1                     4.1                    22161.0        21096.0   
2                     4.3                    69379.0        65881.0   
3                     7.6                    11006.0        10154.0   
4                     6.8                     8961.0         8333.0   

   Unemployed_2002  Unemployment_rate_2002  Civilian_labor_force_2003  \
0         124988.0                     5.9                  2128668.0   
1           1065.0                     4.8                    22695.0   
2           3498.0                     5.0                    72598.0   
3            852.0                     7.7                    11019.0   
4            628.0                     7.0                     8871.0   

   Employed_2003  Unemployed_2003  Unemployment_rate_2003  \
0      2001568.0         127100.0                     6.0   
1        21557.0           1138.0                     5.0   
2        69010.0           3588.0                     4.9   
3        10241.0            778.0                     7.1   
4         8346.0            525.0                     5.9   

   Civilian_labor_force_2004  Employed_2004  Unemployed_2004  \
0                  2138306.0      2018783.0         119523.0   
1                    23241.0        22146.0           1095.0   
2                    74843.0        71061.0           3782.0   
3                    10639.0         9884.0            755.0   
4                     8851.0         8375.0            476.0   

   Unemployment_rate_2004  Civilian_labor_force_2005  Employed_2005  \
0                     5.6                  2140356.0      2045234.0   
1                     4.7                    23887.0        22986.0   
2                     5.1                    76608.0        73581.0   
3                     7.1                    10730.0        10114.0   
4                     5.4                     8837.0         8444.0   

   Unemployed_2005  Unemployment_rate_2005  Civilian_labor_force_2006  \
0          95122.0                     4.4                  2170007.0   
1            901.0                     3.8                    24425.0   
2           3027.0                     4.0                    79806.0   
3            616.0                     5.7

## **Save the Dataframe as CSV**

In [154]:
file = os.path.join(processed_data_directory, 'cleaned_UnemploymentMedianIncome.csv')
unemployment_df.to_csv(file, index=False)

## **Verify cleaned_UnemploymentMedianIncome.csv**

In [155]:
file = os.path.join(processed_data_directory, 'cleaned_UnemploymentMedianIncome.csv')
unemployment_df = pd.read_csv(file)
unemployment_df.head()

FIPS_Code State   county  is_state  Rural_Urban_Continuum_Code_2013  \
0       1000    AL  Alabama      True                             -1.0   
1       1001    AL  Autauga     False                              2.0   
2       1003    AL  Baldwin     False                              3.0   
3       1005    AL  Barbour     False                              6.0   
4       1007    AL     Bibb     False                              1.0   

   Urban_Influence_Code_2013  Metro_2013  Civilian_labor_force_2000  \
0                       -1.0        -1.0                  2147173.0   
1                        2.0         1.0                    21861.0   
2                        2.0         1.0                    69979.0   
3                        6.0         0.0                    11449.0   
4                        1.0         1.0                     8623.0   

   Employed_2000  Unemployed_2000  Unemployment_rate_2000  \
0      2047731.0          99442.0                     4.6   
1        20971.0            890.0                     4.1   
2        67370.0           2609.0                     3.7   
3        10812.0            637.0                     5.6   
4         8160.0            463.0                     5.4   

   Civilian_labor_force_2001  Employed_2001  Unemployed_2001  \
0                  2128027.0      2017467.0         110560.0   
1                    22081.0        21166.0            915.0   
2                    69569.0        66545.0           3024.0   
3                    11324.0        10468.0            856.0   
4                     9134.0         8513.0            621.0   

   Unemployment_rate_2001  Civilian_labor_force_2002  Employed_2002  \
0                     5.2                  2112621.0      1987633.0   
1                     4.1                    22161.0        21096.0   
2                     4.3                    69379.0        65881.0   
3                     7.6                    11006.0        10154.0   
4                     6.8                     8961.0         8333.0   

   Unemployed_2002  Unemployment_rate_2002  Civilian_labor_force_2003  \
0         124988.0                     5.9                  2128668.0   
1           1065.0                     4.8                    22695.0   
2           3498.0                     5.0                    72598.0   
3            852.0                     7.7                    11019.0   
4            628.0                     7.0                     8871.0   

   Employed_2003  Unemployed_2003  Unemployment_rate_2003  \
0      2001568.0         127100.0                     6.0   
1        21557.0           1138.0                     5.0   
2        69010.0           3588.0                     4.9   
3        10241.0            778.0                     7.1   
4         8346.0            525.0                     5.9   

   Civilian_labor_force_2004  Employed_2004  Unemployed_2004  \
0                  2138306.0      2018783.0         119523.0   
1                    23241.0        22146.0           1095.0   
2                    74843.0        71061.0           3782.0   
3                    10639.0         9884.0            755.0   
4                     8851.0         8375.0            476.0   

   Unemployment_rate_2004  Civilian_labor_force_2005  Employed_2005  \
0                     5.6                  2140356.0      2045234.0   
1                     4.7                    23887.0        22986.0   
2                     5.1                    76608.0        73581.0   
3                     7.1                    10730.0        10114.0   
4                     5.4                     8837.0         8444.0   

   Unemployed_2005  Unemployment_rate_2005  Civilian_labor_force_2006  \
0          95122.0                     4.4                  2170007.0   
1            901.0                     3.8                    24425.0   
2           3027.0                     4.0                    79806.0   
3            616.0                     5.7